### PR TITLE
1529 school daily mail - still a bit WIP

### PR DESCRIPTION
**Pull request by tonyheadford**
### Context
[Trello card](https://trello.com/c/t1MF2dg6/1529-create-school-daily-mail-opt-in-pages)
Add Daily Mail "wizard" for school and responsible body users to opt in their school(s)

### Changes proposed in this pull request
Adds the school user journey to opt in their school
Adds the RB journey to opt in all or some of their centrally managed schools

### Guidance to review
Set the `donated_devices` feature flag
As a user of a devolved school, you should see the "Opt in to the Daily Mail’s donated devices scheme" link on the home page.  Follow this to work through the wizard.
As the user of a responsible body with centrally managed schools, You'll see a bullet point under the "Get laptops and tablets" link referring to the Daily Mail scheme. Follow the "Get laptops and tablets" link the the "Opt in to the Daily Mail’s donated devices scheme" link will be seen.

TODO: More specs need adding, policy updates are being done now, some screen shots

### CODE REVIEW

File: app/controllers/responsible_body/donated_devices/interest_controller.rb [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+  #                                        opt_in_choice: 'all_schools',
+  #                                        status: 'incomplete')
+  #   else
+  #     ddr.update!(schools: school_ids,
+  #                 user: current_user,
+  #                 opt_in_choice: 'all_schools',
+  #                 status: 'incomplete')
+  #   end
+  #   ddr
+  # end
```
Comment (aldavidson) [inline app/controllers/responsible_body/donated_devices/interest_controller.rb:167]:
```suggestion
```

File: app/controllers/responsible_body/donated_devices/interest_controller.rb [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    parms = donated_device_params.merge(responsible_body: @responsible_body,
+                                        user: current_user)
+    DonatedDeviceRequest.new(parms)
+  end
+
+  def find_request!
+    ddr = DonatedDeviceRequest.for_responsible_body(@responsible_body).first
+    if ddr.nil?
+      # send backto beginning if they've got here without a request being saved
+      flash[:error] = 'They was a problem with your request'
```
Comment (aldavidson) [inline app/controllers/responsible_body/donated_devices/interest_controller.rb:185]:
```suggestion
      flash[:error] = 'There was a problem with your request'
```

File: app/controllers/school/donated_devices/interest_controller.rb [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    ddr = DonatedDeviceRequest.incomplete.for_school(@school).first
+    ddr = build_donated_device_request if ddr.nil?
+    @request = present(ddr)
+  end
+
+  def find_request!
+    ddr = DonatedDeviceRequest.for_school(@school).first
+    if ddr.nil?
+      # send backto beginning if they've got here without a request being saved
+      flash[:error] = 'They was a problem with your request'
```
Comment (aldavidson) [inline app/controllers/school/donated_devices/interest_controller.rb:97]:
```suggestion
      flash[:error] = 'There was a problem with your request'
```

File: app/models/school.rb [hunk 4]
```diff
@@ -18,7 +18,7 @@ class School < ApplicationRecord
   has_many :devices_ordered_updates, class_name: 'Computacenter::DevicesOrderedUpdate',
                                      primary_key: :computacenter_reference,
                                      foreign_key: :ship_to
-  has_one :donated_device_request
+  # has_one :donated_device_request
```
Comment (aldavidson) [inline app/models/school.rb:21]:
```suggestion
```

File: app/views/responsible_body/donated_devices/interest/about.html.erb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    <h1 class="govuk-heading-xl"><%= title %></h1>
+
+    <h2 class="govuk-heading-m">Devices can vary</h2>
+    <p class="govuk-body">A variety of devices are being donated. The devices a school or college could get might not all be the same.</p>
+
+    <p class="govuk-body">Every device will meet the <%= govuk_link_to 'minimum requirements needed for remote learning', 'https://www.computacenter.com/uk/daily-mail', target: '_blank', rel: 'noopener noreferrer' %>.</p>
+    
+    <h2 class="govuk-heading-m">Configure at your own cost</h2>
+    <p class="govuk-body">Devices are provided with factory settings, schools and colleges will need to configure them at their own cost before lending them to children and young people.</p>
+    <p class="govuk-body govuk-!-margin-bottom-7">Schools and colleges will need to purchase licences for any Google Chromebooks they receive.</p>
```
Comment (aldavidson) [inline app/views/responsible_body/donated_devices/interest/about.html.erb:16]: This seems a little odd - I would have expected to purchase licenses for any Windows laptops I receive, but not for Chromebooks. Are we sure that's right?

File: app/views/school/donated_devices/interest/about.html.erb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    <h2 class="govuk-heading-m">Devices can vary</h2>
+    <p class="govuk-body">A variety of devices are being donated. The devices you could get might not be the same.</p>
+
+    <p class="govuk-body">Every device will meet the <%= govuk_link_to 'minimum requirements needed for remote learning', 'https://www.computacenter.com/uk/daily-mail', target: '_blank', rel: 'noopener noreferrer' %>.</p>
+    <h2 class="govuk-heading-m">
+      Configure at your own cost
+    </h2>
+  
+    <p class="govuk-body">Devices are provided with factory settings and you will need to configure them at your own cost before lending them to children and young people.</p>
+    <p class="govuk-body govuk-!-margin-bottom-7">You will need to purchase licences for any Google Chromebooks you receive.</p>
```
Comment (aldavidson) [inline app/views/school/donated_devices/interest/about.html.erb:19]: As above

File: spec/features/responsible_body/donated_devices_spec.rb [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+  let(:school) { create(:school, :with_preorder_information, responsible_body: responsible_body) }
+
+  before do
+    responsible_body.update! who_will_order_devices: 'responsible_body'
+    school.preorder_information.update!(who_will_order_devices: 'responsible_body')
+    sign_in_as user
+  end
+
+  scenario 'RB that centrally manages schools can navigate to donated device form from the home page' do
+    given_centrally_managed_school
```
Comment (aldavidson) [inline spec/features/responsible_body/donated_devices_spec.rb:15]:
```suggestion
    given_i_have_a_centrally_managed_school
```

File: spec/features/responsible_body/donated_devices_spec.rb [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    and_then_i_continue_to_the_next_page
+
+    then_i_am_asked_if_i_am_still_interested
+    and_i_indicate_that_i_am_still_interested
+
+    # Now at: Which schools do you want to opt in?
+  end
+
+  scenario 'RB navigating to donated device form but declining interest' do
+    given_centrally_managed_school
```
Comment (aldavidson) [inline spec/features/responsible_body/donated_devices_spec.rb:36]:
```suggestion
    given_i_have_a_centrally_managed_school
```

File: spec/features/responsible_body/donated_devices_spec.rb [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+    and_i_click_the_donated_devices_link
+    then_i_am_asked_if_i_am_interested_in_devices
+    and_i_indicate_that_i_am_not_interested
+
+    then_i_see_that_i_have_not_been_opted_in
+  end
+
+  scenario 'RB that has devolved schools cannot see the donated device form from the home page' do
+    given_devolved_school
```
Comment (aldavidson) [inline spec/features/responsible_body/donated_devices_spec.rb:48]:
```suggestion
    given_i_have_a_devolved_school
```

File: spec/features/responsible_body/donated_devices_spec.rb [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    and_i_navigate_to_the_devices_page
+    then_see_that_i_cannot_opt_in_my_schools
+  end
+
+  scenario 'RB that centrally managed but cannot see the donated device form from the home page without feature flag', with_feature_flags: { donated_devices: 'inactive' } do
+    given_centrally_managed_school
```
Comment (aldavidson) [inline spec/features/responsible_body/donated_devices_spec.rb:54]:
```suggestion
    given_i_have_a_centrally_managed_school
```

File: spec/features/responsible_body/donated_devices_spec.rb [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+    and_i_navigate_to_the_devices_page
+    then_see_that_i_cannot_opt_in_my_schools
+  end
+
+private
+
+  def given_devolved_school
```
Comment (aldavidson) [inline spec/features/responsible_body/donated_devices_spec.rb:61]:
```suggestion
  def given_i_have_a_devolved_school
```

File: spec/features/responsible_body/donated_devices_spec.rb [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    school.preorder_information.update!(who_will_order_devices: 'school')
+  end
+
+  def given_centrally_managed_school
```
Comment (aldavidson) [inline spec/features/responsible_body/donated_devices_spec.rb:65]:
```suggestion
  def given_i_have_a_centrally_managed_school
```

File: spec/features/responsible_body/donated_devices_spec.rb [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+  def and_i_indicate_that_i_am_still_interested
+    choose 'Yes'
+    click_on 'Continue'
+  end
+
+  def and_then_i_continue_to_the_next_page
+    click_on 'Continue'
+  end
+
+  def then_see_that_i_can_opt_in_my_schools
```
Comment (aldavidson) [inline spec/features/responsible_body/donated_devices_spec.rb:97]:
```suggestion
  def then_i_see_that_i_can_opt_in_my_schools
```

File: spec/features/responsible_body/donated_devices_spec.rb [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+    expect(page).to have_content('Opt in to the Daily Mail’s donated devices scheme')
+  end
+
+  def then_see_that_i_cannot_opt_in_my_schools
```
Comment (aldavidson) [inline spec/features/responsible_body/donated_devices_spec.rb:101]:
```suggestion
  def then_i_see_that_i_cannot_opt_in_my_schools
```

File: spec/features/responsible_body/donated_devices_spec.rb [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+  scenario 'RB that has devolved schools cannot see the donated device form from the home page' do
+    given_devolved_school
+    and_i_navigate_to_the_devices_page
+    then_see_that_i_cannot_opt_in_my_schools
+  end
+
+  scenario 'RB that centrally managed but cannot see the donated device form from the home page without feature flag', with_feature_flags: { donated_devices: 'inactive' } do
+    given_centrally_managed_school
+    and_i_navigate_to_the_devices_page
+    then_see_that_i_cannot_opt_in_my_schools
```
Comment (aldavidson) [inline spec/features/responsible_body/donated_devices_spec.rb:56]:
```suggestion
    then_i_see_that_i_cannot_opt_in_my_schools
```

File: spec/features/responsible_body/donated_devices_spec.rb [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+    then_i_am_asked_if_i_am_interested_in_devices
+    and_i_indicate_that_i_am_not_interested
+
+    then_i_see_that_i_have_not_been_opted_in
+  end
+
+  scenario 'RB that has devolved schools cannot see the donated device form from the home page' do
+    given_devolved_school
+    and_i_navigate_to_the_devices_page
+    then_see_that_i_cannot_opt_in_my_schools
```
Comment (aldavidson) [inline spec/features/responsible_body/donated_devices_spec.rb:50]:
```suggestion
    then_i_see_that_i_cannot_opt_in_my_schools
```

File: spec/features/school/donated_devices_spec.rb [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+  let(:user) { create(:school_user) }
+  let(:school) { user.school }
+
+  before do
+    school.create_preorder_information!(who_will_order_devices: 'school')
+    sign_in_as user
+  end
+
+  scenario 'devolved school can navigate to donated device form from the home page' do
+    given_devolved_school
```
Comment (aldavidson) [inline spec/features/school/donated_devices_spec.rb:13]:
```suggestion
    given_i_have_a_devolved_school
```

File: spec/features/school/donated_devices_spec.rb [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    and_i_navigate_to_the_home_page
+    then_see_that_i_can_opt_in
```
Comment (aldavidson) [inline spec/features/school/donated_devices_spec.rb:15]:
```suggestion
    then_i_see_that_i_can_opt_in
```

File: spec/features/school/donated_devices_spec.rb [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  scenario 'centrally managed school cannot see the donated device form from the home page' do
+    given_centrally_managed_school
```
Comment (aldavidson) [inline spec/features/school/donated_devices_spec.rb:19]:
```suggestion
    given_i_have_a_centrally_managed_school
```

File: spec/features/school/donated_devices_spec.rb [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+    and_i_navigate_to_the_home_page
+    then_see_that_i_cannot_opt_in
```
Comment (aldavidson) [inline spec/features/school/donated_devices_spec.rb:21]:
```suggestion
    then_i_see_that_i_cannot_opt_in
```

File: spec/features/school/donated_devices_spec.rb [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  scenario 'school cannot see donated device form from homepage without feature flag', with_feature_flags: { donated_devices: 'inactive' } do
+    given_centrally_managed_school
```
Comment (aldavidson) [inline spec/features/school/donated_devices_spec.rb:25]:
```suggestion
    given_i_have_a_centrally_managed_school
```

File: spec/features/school/donated_devices_spec.rb [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+    and_i_navigate_to_the_home_page
+    then_see_that_i_cannot_opt_in
```
Comment (aldavidson) [inline spec/features/school/donated_devices_spec.rb:27]:
```suggestion
    then_i_see_that_i_cannot_opt_in
```

File: spec/features/school/donated_devices_spec.rb [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+private
+
+  def given_devolved_school
```
Comment (aldavidson) [inline spec/features/school/donated_devices_spec.rb:32]:
```suggestion
  def given_i_have_a_devolved_school
```

File: spec/features/school/donated_devices_spec.rb [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+    school.preorder_information.update!(who_will_order_devices: 'school')
+  end
+
+  def given_centrally_managed_school
```
Comment (aldavidson) [inline spec/features/school/donated_devices_spec.rb:36]:
```suggestion
  def given_i_have_a_centrally_managed_school
```

File: spec/features/school/donated_devices_spec.rb [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+    school.preorder_information.update!(who_will_order_devices: 'responsible_body')
+  end
+
+  def and_i_navigate_to_the_home_page
+    visit home_school_path(school)
+  end
+
+  def then_see_that_i_can_opt_in
```
Comment (aldavidson) [inline spec/features/school/donated_devices_spec.rb:44]:
```suggestion
  def then_i_see_that_i_can_opt_in
```

File: spec/features/school/donated_devices_spec.rb [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+    expect(page).to have_content('Opt in to the Daily Mail’s donated devices scheme')
+  end
+
+  def then_see_that_i_cannot_opt_in
```
Comment (aldavidson) [inline spec/features/school/donated_devices_spec.rb:48]:
```suggestion
  def then_i_see_that_i_cannot_opt_in
```

File: app/controllers/responsible_body/donated_devices/interest_controller.rb [hunk 27]
```diff
@@ -0,0 +1,227 @@
...
+        else
+          render :not_interested
+        end
+      else
+        render :interest_confirmation, status: :unprocessable_entity
+      end
+    end
+  end
+
+  def all_or_some_schools
```
Comment (aldavidson) [inline app/controllers/responsible_body/donated_devices/interest_controller.rb:42]: This method is pretty long, and a bit tricky to follow all the ifs/elses - could some of the cases be factored out to a separate method?

File: app/controllers/responsible_body/donated_devices/interest_controller.rb [hunk 28]
```diff
@@ -0,0 +1,227 @@
...
+          @request.save!
+          redirect_to responsible_body_donated_devices_select_schools_path
+        end
+      else
+        render :all_or_some_schools, status: :unprocessable_entity
+      end
+    end
+  end
+
+  def select_schools
```
Comment (aldavidson) [inline app/controllers/responsible_body/donated_devices/interest_controller.rb:68]: Ditto - this method is even longer than the previous one, and some of the logic seems common between the two - could some of the cases be factored out to a separate method?

I'm thinking of simple things like:

```
          if @request.complete?
            redirect_to responsible_body_donated_devices_opted_in_path
          else
            redirect_to responsible_body_donated_devices_what_devices_do_you_want_path
          end
```

could be refactored to:

```
redirect_to next_url_for(@request)

(...)

def next_url_for(request: @request)
  request.complete? responsible_body_donated_devices_opted_in_path : responsible_body_donated_devices_what_devices_do_you_want_path
end
```
etc

**Review by aldavidson — APPROVED** [review]
This is huge amounts of work, and it seems like I've raised loads of issues - but they're all stylistic points and none of them are blockers if releasing this is urgent. Good work! :100:

File: app/services/donated_device_requests_exporter.rb [hunk 29]
```diff
@@ -40,22 +40,25 @@ def self.headings
...
+          request.device_types.join(','),
+          request.units,
+        ]
+      end
     end
   end
 
   def donated_device_requests
-    DonatedDeviceRequest.all.includes(:user, school: :responsible_body).order(:asc)
+    DonatedDeviceRequest.complete.includes(:user, :responsible_body).order(:asc)
```
Comment (asmega) [inline app/services/donated_device_requests_exporter.rb:62]:
```
[12] pry(main)> User.order(:asc).to_sql
=> "SELECT \"users\".* FROM \"users\" ORDER BY \"asc\" ASC"
```
I think you want `order(id: :asc)`

File: app/services/donated_device_requests_exporter.rb [hunk 30]
```diff
@@ -40,22 +40,25 @@ def self.headings
...
-        request.school.urn,
-        request.school.computacenter_reference,
-        request.school.responsible_body.computacenter_reference,
-        request.user.full_name,
-        request.user.email_address,
-        request.user.telephone,
-        request.device_types.join(','),
-        request.units,
-      ]
+      request.schools.each do |school_id|
```
Comment (asmega) [inline app/services/donated_device_requests_exporter.rb:43]: this should be fine if numbers are small but could pose a bottleneck

doing `schools = School.where(id: request.schools)` is marginally better but still means for every request a query is being made

File: app/services/donated_device_requests_exporter.rb (hunk 29 above)
Comment (tonyheadford) [reply app/services/donated_device_requests_exporter.rb:62]: That's fixed already in my branch - not sure what's going on there?

File: app/services/donated_device_requests_exporter.rb [hunk 31]
```diff
@@ -40,22 +40,25 @@ def self.headings
...
+        school = School.find(school_id)
```
Comment (asmega) [inline app/services/donated_device_requests_exporter.rb:44]: you want to eager load rb on the school

File: app/services/donated_device_requests_exporter.rb (hunk 29 above)
Comment (asmega) [inline app/services/donated_device_requests_exporter.rb:62]: the eager load of rb on request is never used from what i can tell so can delete

File: db/schema.rb [hunk 32]
```diff
@@ -399,6 +404,7 @@
     t.boolean "orders_devices"
     t.datetime "techsource_account_confirmed_at"
     t.datetime "deleted_at"
+    t.boolean "rb_level_access", default: false, null: false
```
Comment (asmega) [inline db/schema.rb:407]: change was introduced elsewhere can we delete this line please

File: db/schema.rb [hunk 33]
```diff
@@ -106,6 +106,8 @@
     t.text "original_cc_ship_to_number"
     t.datetime "cc_import_api_timestamp"
     t.string "cc_import_api_transaction_id"
+    t.boolean "cc_rb_user"
+    t.boolean "original_cc_rb_user"
```
Comment (asmega) [inline db/schema.rb:110]: change was introduced elsewhere can we delete these lines please

Comment (tonyheadford) [reply db/schema.rb:110]: Good spot

File: db/schema.rb (hunk 32 above)
Comment (tonyheadford) [reply db/schema.rb]: Good spot but that one should be in main (and have just rebased)

**Pull request merged**